openedx/xblock-lti-consumer
# Proctoring Service Implementation In-Browser Proctoring Launch

**Description:**

This pull request introduces support for the LTI 1.3 Proctoring Services specification. It includes support for the in-browser proctoring launch flow. It does not currently include support for the proctoring services.

The previous iteration of these changes can be found [here](https://github.com/openedx/xblock-lti-consumer/pull/297). This pull request takes its place.

**Changes:**

Here is a digest of important changes.

* An `Lti1p3ProctoringLaunchData` data class is introduced. It should be included as an attribute of the `Lti1p3LaunchData` data class to provide necessary proctoring data for a launch.
* An `LtiProctoringConsumer` class is introduced. This class is used to generate LTI proctoring launch requests and to decode and validate the JWT send back by the Tool with the `LtiStartAssessment` message.
* An `lti_1p3_proctoring_enabled` BooleanField is added to the `LtiConfiguration` model. This field controls whether proctoring is enabled for a particular LTI integration.
* The `launch_gate_endpoint` has been modified to support `LtiStartProctoring` and `LtiEndAssessment` LTI launch messages.
* A `start_proctoring_assessment_endpoint` is added to support `LtiStartAssessment` messages from the Tool.
* An `LTI_1P3_PROCTORING_ASSESSMENT_STARTED` signal is introduced. This signal is emitted when the `LtiStartAssessment` message is sent from the Tool to inform users of the library that the `LtiStartAssessment` message has been received.

**Testing Instructions:**

You will need a user of this new code. You can the edx-exams repository to do so. You can use this pull request ().

* In your local checkout of edx-exams, set the following settings in `edx_exams/settings/private.py`.

  * ROOT_URL = <NGROK_URL>
  * LMS_ROOT_URL = <NGROK_URL>
  * SESSION_COOKIE_SAMESITE = "None"
  * SESSION_COOKIE_SECURE = True
  * ALLOWED_HOSTS = ['*']

* The requirements files should be pointing at the correct branch of xblock-lti-consumer and should install it correctly. If there are problems, check out the xblock-lti-consumer library locally and pip install it. If the proposed Proctoring Services implementation is still on a branch, check out the branch an install it.

* Log in to the edx-exams service using the ngrok URL - `<NGROK_URL>/admin`. The reason it's important to log in via the ngrok URL is to ensure that the session cookie is set on the proper domain so that the cookie is included by the browser when the Tool makes requests to the NGROK_URL. If you log in through localhost, then when the Tool makes requests to the NGROK_URL, the user will be the AnonymousUser. This will cause the flow to fail.
* Set up the [Proctoring Certification Suite 1.0](https://certification.imsglobal.org/certification/proctoring/launch.html).
  * Create a new `LtiConfiguration` instance in the edx-exams admin panel.

    * Set the "Config store" to "Configuration Stored on this model".
  
  * The process of setting this testing suite up is very similar to the [existing instructions](https://github.com/openedx/xblock-lti-consumer#lti-13) for testing LTI 1.3 launches in the xblock-lti-consumer README. Use that as a reference. Instead of setting the data on the XBlock, set it on the LtiConfiguration in the admin panel. Instead of setting the data in the LTI 1.3 launch testing tool, set it in the Proctoring Certification Suite 1.0.

* In your browser, hit the following URL to start the LTI proctoring flow: `<NGROK_URL>/lti/start_proctoring/1`. 1 represents the database ID for the `LtiConfiguration`, so set it appropriately if you have more than one `LtiConfiguration` you're testing with.

**Quality Assurance:**

This code has been tested manually with a `CONFIG_EXTERNAL` configuration using the aforementioned edx-exams code.

Here are screenshots from the Proctoring Services test suite published by 1EdTech.

Successful LTI Launch with "LtiStartProctoring" message:

<img width="1028" alt="image" src="https://user-images.githubusercontent.com/11871801/200682726-48f91b8a-fe49-4f3e-8515-9f4da322d52c.png">

There are a number of tests under the "Start Assessment" section. They're shown below.

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/11871801/200685709-7d7f1b44-aa5c-4b45-a0a8-3f785f396656.png">


All tests intended to test failures show this page.

<img width="697" alt="image" src="https://user-images.githubusercontent.com/11871801/200685672-5d09e929-d60d-4d1e-9e6a-e508b08b8649.png">


The last test, "Valid LTI Message", shows this page.

<img width="680" alt="image" src="https://user-images.githubusercontent.com/11871801/200685799-c3a1cc11-d8b3-4a98-ac95-0073838e33b0.png">
